Enable KAS

antt1995 opened the pull request:
Enabling KAS

Not sure what to put or where in the Readme

antt1995 commented:
https://github.com/sameersbn/docker-gitlab/issues/2494#issue-1107359331

wulfdj commented:
Are these settings all I need? I've tried to change on the running container but still doesn't work. On the clusters page I get the error message that GRPC is not available

wulfdj commented:
Are these settings all I need? I've tried to change on the running container but still doesn't work. On the clusters page I get the error message that GRPC is not available

blattersturm commented:
This *seems* to only set the configuration settings in case you're trying to use an [external KAS install](https://docs.gitlab.com/ee/administration/clusters/kas.html#use-an-external-installation).

The Omnibus repo seems to use [some more](https://gitlab.com/gitlab-org/omnibus-gitlab/-/blob/63e13f6a9b3489969e790dc30aeb3e83348ee577/config/software/gitlab-kas.rb) steps to deploy the built-in KAS instance.

kkimurak reviewed (changes requested):
@antt1995 Thank you for your contribution.  
Those settings seems fine, but we need some step to make them accessible.

1. Please add function (like `gitlab_configure_gitlab_kas()`) to `assets/runtime/functions`  
  Maybe simplest reference is `gitlab_configure_monitoring()`:
https://github.com/sameersbn/docker-gitlab/blob/b34d48af65fc002000c84c4d71460ac9426af1ab/assets/runtime/functions#L264-L273
  So the implementation will be like:
    ````sh
    gitlab_configure_gitlab_kas() {
      echo "Configuring gitlab::KAS..."
    
      update_template ${GITLAB_CONFIG} \
        GITLAB_KAS_ENABLED \
        GITLAB_KAS_SECRET \
        GITLAB_KAS_EXTERNAL \
        GITLAB_KAS_INTERNAL \
        GITLAB_KAS_PROXY
    }
    ````

2. Call the function in `configure_gitlab()`. This function will be called on container startup.
https://github.com/sameersbn/docker-gitlab/blob/b34d48af65fc002000c84c4d71460ac9426af1ab/assets/runtime/functions#L1814-L1877

3. (optional) The variable naming rule is not strictly defined, but it seems  converting all yaml key name into UPPER_CASE is common approach for this repository. 
    - GITLAB_KAS_SECRET -> GITLAB_KAS_SECRET_FILE
    - GITLAB_KAS_EXTERNAL -> GITLAB_KAS_EXTERNAL_URL
    - GITLAB_KAS_INTERNAL -> GITLAB_KAS_INTERNAL_URL
    - GITLAB_KAS_PROXY -> GITLAB_KAS_K8S_PROXY_URL

antt1995 commented:
@kkimurak
I have added that to the Functions

kkimurak reviewed (changes requested):
Passed build and test! Thanks for your contribution.  

Sorry for forgetting to check in the last review, but could you please fill these additional tasks?

- add a description of the variables you added to the ["Available Configuration Parameters" section of the README.md](https://github.com/sameersbn/docker-gitlab#available-configuration-parameters)
- check default values for the variables

antt1995 commented:
Building Now with basic info in readme

antt1995 commented:
> Not sure what to put or where in the Readme

Still not sure what to put but have put the vars and their defaults

kkimurak commented on README.md:
```diff
@@ -776,6 +777,16 @@ These options should contain something like:
 - `PIWIK_URL=piwik.example.org`
 - `PIWIK_SITE_ID=42`
 
+#### Gitlab KAS
+
+Basic Gitlab KAS Settings 
+
+- `GITLAB_KAS_ENABLED=${GITLAB_KAS_ENABLED:-false}`
+- `GITLAB_KAS_SECRET=${GITLAB_INSTALL_DIR}/.gitlab_kas_secret}`
+- `GITLAB_KAS_EXTERNAL="wss://kas.example.com"}`
+- `GITLAB_KAS_INTERNAL="grpc://localhost:8153"}`
+- `GITLAB_KAS_PROXY=`
+
```
I'm very sorry that I couldn't get enough time to explain it enough, but I hope you read this patch (and the result of applying it) to understand what you should have written in the README..

````patch
diff --git a/README.md b/README.md
index f8862af3..92e12bc7 100644
--- a/README.md
+++ b/README.md
@@ -777,16 +777,6 @@ These options should contain something like:
 - `PIWIK_URL=piwik.example.org`
 - `PIWIK_SITE_ID=42`
 
-#### Gitlab KAS
-
-Basic Gitlab KAS Settings 
-
-- `GITLAB_KAS_ENABLED=${GITLAB_KAS_ENABLED:-false}`
-- `GITLAB_KAS_SECRET=${GITLAB_INSTALL_DIR}/.gitlab_kas_secret}`
-- `GITLAB_KAS_EXTERNAL="wss://kas.example.com"}`
-- `GITLAB_KAS_INTERNAL="grpc://localhost:8153"}`
-- `GITLAB_KAS_PROXY=`
-
 #### Available Configuration Parameters
 
 *Please refer the docker run command options for the `--env-file` flag where you can specify all required environment variables in a single file. This will save you from writing a potentially long docker run command. Alternatively you can use docker-compose. docker-compose users and Docker Swarm mode users can also use the [secrets and config file options](#docker-secrets-and-configs)*
@@ -1093,6 +1083,26 @@ Default Google key file. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_JSO
 
 Cron notation for the GitLab pipeline schedule worker. Defaults to `'19 * * * *'`
 
+##### `GITLAB_KAS_ENABLED`
+
+Enable/Disable GitLab agent server for Kubernetes (KAS). See details on [official documentation](https://docs.gitlab.com/ee/administration/clusters/kas.html). Defaults to `false`
+
+##### `GITLAB_KAS_SECRET`
+
+File that contains the secret key for verifying access for GitLab KAS. Defaults to `${GITLAB_INSTALL_DIR}/.gitlab_kas_secret`
+
+##### `GITLAB_KAS_EXTERNAL`
+
+User-facing URL for the in-cluster agent. Defaults to `"wss://kas.example.com"`
+
+##### `GITLAB_KAS_INTERNAL`
+
+Internal URL for the GitLab backend. Defaults to `"grpc://localhost:8153"`
+
+##### `GITLAB_KAS_PROXY`
+
+The URL to the Kubernetes API proxy (used by GitLab users). No default.
+
 ##### `GITLAB_LFS_ENABLED`
 
 Enable/Disable Git LFS support. Defaults to `true`.
````

kkimurak replied:
By the way, the official documentation describes that [you need to set an secret api key to `gitlab_kas['api_secret_key']` like:](https://docs.gitlab.com/ee/administration/clusters/kas.html#use-an-external-installation)

````ruby
gitlab_kas['api_secret_key'] = 'Your shared secret between GitLab and KAS'
````

Do you know anything about it? I have never used KAS and was interested in how it should be set from the user's point of view.

antt1995 replied:
I have made the suggested changes to the Readme,
I have no idea about the gitlab_kas['api_secret_key']

kkimurak replied:
Oh, sorry, I have sent you in a diff format. The "+" sign doesn't need to be included in the README.md. Just write like:

```md
##### `GITLAB_KAS_ENABLED`

Enable/Disable GitLab agent server for Kubernetes (KAS). See details on [official documentation](https://docs.gitlab.com/ee/administration/clusters/kas.html). Defaults to `false`

##### `GITLAB_KAS_SECRET`

File that contains the secret key for verifying access for GitLab KAS. Defaults to `${GITLAB_INSTALL_DIR}/.gitlab_kas_secret`

##### `GITLAB_KAS_EXTERNAL`

User-facing URL for the in-cluster agent. Defaults to `"wss://kas.example.com"`

##### `GITLAB_KAS_INTERNAL`

Internal URL for the GitLab backend. Defaults to `"grpc://localhost:8153"`

##### `GITLAB_KAS_PROXY`

The URL to the Kubernetes API proxy (used by GitLab users). No default.

```

Anyway, thank you for answering about the API key. I hope someone has some knowledge of KAS and can see if it works ...

antt1995 replied:
Updated to remove +

kkimurak replied:
Sorry for the many corrections, but as @sachilles says, this document looks inaccurate. I told you to remove it in the last review, but if you want to keep it, I think we should fix it and write a little more detail.

```suggestion
#### Gitlab KAS

GitLab agent server for Kubernetes (KAS) is disabled by default, but you can enable it by setting configuration parameter [`GITLAB_KAS_ENABLED`](#GITLAB_KAS_ENABLED) to true.
You can specify the internal/external URL for KAS via [`GITLAB_KAS_INTERNAL`](#GITLAB_KAS_INTERNAL) and [`GITLAB_KAS_EXTERNAL`](#GITLAB_KAS_EXTERNAL). If you set up proxy URL, use `GITLAB_KAS_PROXY`.

See official documentation : https://docs.gitlab.com/ee/administration/clusters/kas.html
```

kkimurak commented on assets/runtime/env-defaults:
```diff
@@ -604,3 +604,10 @@ GITLAB_CONTENT_SECURITY_POLICY_DIRECTIVES_SCRIPT_SRC=${GITLAB_CONTENT_SECURITY_P
+
+## Gitlab KAS
+GITLAB_KAS_ENABLED=${GITLAB_KAS_ENABLED:-false}
+GITLAB_KAS_SECRET=${GITLAB_KAS_SECRET:-${GITLAB_INSTALL_DIR}/.gitlab_kas_secret}}
```
It seems that I have send you a suggestion contains typo. Sorry again and could you please fix this like:

```suggestion
GITLAB_KAS_SECRET=${GITLAB_KAS_SECRET:-${GITLAB_INSTALL_DIR}/.gitlab_kas_secret}
```

kkimurak reviewed (changes requested):
Sorry for late, I have added review for README and fixed typo in my previous suggestion. Could you confirm them?

Best,

kkimurak commented:
@antt1995 Do you mean you confirmed KAS works as expected applying this PR? Then I can approve. Thank you for your contribution and sorry for my rough review.

By the way I have no permission to merge. Please ask for maintainer (`@sachilles` is active, I think)

antt1995 commented:
@kkimurak 
> Sorry for late, I have added review for README and fixed typo in my previous suggestion. Could you confirm them?

That's what I was Confirming

kkimurak commented:
@antt1995 ok Thank you. I approve the code changes themselves, but I'm not sure if KAS works. 
I would like to find time to see if kas works. As mentioned above, using the internal KAS should require more work. If you are ready to use external KAS, please ask `@sachilles` to merge this (It is nice to confirm if it works building image by youself).

tiehfood commented:
Would love this feature to be included

sachilles commented:
Hi there, sorry for being late. I guess @kkimurak did a great job with his valuable hints. However, I don't use KAS and thus, I can't confirm that this PR works.

Nevertheless, I'd like to ask if the configuration file for secrets (`GITLAB_KAS_SECRET=${GITLAB_INSTALL_DIR}/.gitlab_kas_secret}`) will be generated dynamically? If not, I guess the README.md has to be extended in order to explain the configuration for KAS in detail.

kkimurak commented on README.md:
```diff
@@ -51,6 +51,7 @@
     - [Host UID / GID Mapping](#host-uid--gid-mapping)
     - [Piwik](#piwik)
     - [Exposing ssh port in dockerized gitlab-ce](docs/exposing-ssh-port.md)
+    - [Gitlab KAS](#Gitlab KAS)
```
White space in id will be converted to hyphen (`-`).

```suggestion
    - [Gitlab KAS](#Gitlab-KAS)
```

kkimurak reviewed (commented):
Hi @antt1995 , orry for the tedious review process, but after being pointed out by @sachilles I reviewed the README again. Could you please confirm it?

edit: Oops sorry but [I have added comment to "resolved" thread](https://github.com/sameersbn/docker-gitlab/pull/2598#discussion_r932165717) (and hidden by default). I apologize for the inconvenience, but for the second item, press "show resolved" and see the last comment.

eslamhossam23 commented:
Hello @antt1995 , I tested this pull request on my Gitlab instance but I haven't been able to make it work.

The KAS was detected as installed by the UI as shown here:
![image](https://user-images.githubusercontent.com/23102450/184830068-2fa2a297-b8d0-49fd-90b8-26fc692bfb21.png)

However, when adding a new agent, I always got the error message: `GRPC::Unavailable`.
Upon inspecting the container via `ss -lp`, I also noticed that the KAS service was never started at port 8153 as it should have been.

Should the service start by itself automatically by setting `GITLAB_KAS_ENABLED` to `true` or is there something that needs to be done after that?

antt1995 commented:
Did you map the port in your docker config "-p 8153:8153" or in docker-compose "- 8153:8153"

eslamhossam23 commented:
Hello @antt1995 yes, I had the port correctly mapped.

The container itself haven't had started the KAS service inside it.
I checked using `ss -lp` and verified that it was not started.

tiehfood commented:
@eslamhossam23 I deployed the KAS service with separate docker image and set the variables from this PR corresponding. Then it was working for me

kkimurak commented:
@eslamhossam23 As already reported above (https://github.com/sameersbn/docker-gitlab/pull/2598#issuecomment-1172864976), this image currently does not provides built-in KAS. This PR requires external KAS is exists.

If you have time to do that, could you please try to build and test my [`support-builtin-kas` branch](https://github.com/kkimurak/docker-gitlab/tree/support-builtin-kas) ?  
In addition to the work done by @antt1995, I've made sure to build gitlab-kas when building the image and add setting for supervisord to launch kas if enabled. Also I have added configuration file for KAS and minimal configuration parameters. 

I have never used Kubernetes so never tested if the registration success, but made sure the `gitlab-kas` service starts without exiting.  
Here are new entries for env of `docker-compose.yml` on test:

````yaml
services:
  gitlab:
    environment:
      GITLAB_KAS_ENABLED: 'true'
      # If we don't set this, built-in gitlab-kas will exit with an error
      # See https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/blob/8069d9b79763859b5bcead437a28f4dc73b66ae8/cmd/kas/kasapp/configured_app.go#L96-98
      OWN_PRIVATE_API_URL: grpc://127.0.0.1:8155
````

sachilles commented:
Hi,

is this PR considered to be ready to merge?

I guess (and based on this discussion) there should be some comments on using KAS with this image (something like if you want using KAS you'll need to deploy an additional docker image), isn't it?

exploding-dragon commented:
Hi, is this PR considered to be ready to merge?

defp commented:
> @eslamhossam23 I deployed the KAS service with separate docker image and set the variables from this PR corresponding. Then it was working for me

Is there any tutorial? @tiehfood
